**1.How many multiplications and additions do you need to perform a matrix multiplication between a (n, k) and (k, m) matrix? Explain.**

**ANSWER**
To solve this question i have taken example of two matrix A of 2x3 and B of 3x4,
the first row of A which contains 3 elements will be multiplied with first column of B which contains 3 elements, it will contain 3 multiplication operation and 2 addition operations, the same thing will happen for all elements in the resulting matrix, the no. of element in resulting matrix is 2x4,
Hence the total no. of multiplication will be 3x2x4 = 24
the total no. of addition will be 2x2x4 = 16
we can derive the general formula for two matrix a (n, k) and (k, m) will be
Multiplications --> n x m x k
Additions --> n x m x (k-1).

**2.Write Python code to multiply the above two matrices. Solve using list of lists and then use numpy. Compare the timing of both solutions. Which one is faster? Why?**

In [ ]:
#using list os list
%%time
def matrix_multiply(A, B):
     n = len(A)
     k = len(A[0])
     k2 = len(B)
     m = len(B[0])
     if k != k2:
      print('error in dimention')
      return
     result = []
     for i in range(m):
        row = [0] * n
        result.append(row)
     for i in range(n):
        for j in range(m):
            for l in range(k):
                result[i][j] += A[i][l] * B[l][j]

     return result

# Example usage
A = [[1, 2],
    [3, 4]]

B = [[5, 6],
    [7, 8]]

result = matrix_multiply(A, B)
print("Result of A * B:")
for row in result:
    print(row)


Result of A * B:
[19, 22]
[43, 50]
CPU times: user 1.76 ms, sys: 0 ns, total: 1.76 ms
Wall time: 2.32 ms


In [ ]:
#using numpy predefined function
%%time
import numpy as np
A = np.array([[1, 2],
    [3, 4]])

B = np.array([[5, 6],
    [7, 8]
])

result = np.matmul(A, B)

print("Result of A * B:")
print(result)


Result of A * B:
[[19 22]
 [43 50]]
CPU times: user 816 µs, sys: 0 ns, total: 816 µs
Wall time: 944 µs


CPU time in Lisit of list 1.76 ms and time in numpy is 816 µs thus numpy is faster


**3.Finding the highest element in a list requires one pass of the array. Finding the second highest element requires 2 passes of the the array. Using this method, what is the time complexity of finding the median of the array? Can you suggest a better method? Can you implement both these methods in Python and compare against numpy.median routine in terms of time?**

**ANSWER**
using the first method the complexity will be $O(n^2)$ because for finding largest elements it takes $O(n)$ time and repeat it for n elements store all elemts in the new array and just find mid element it will take total $O(n^2)$   

second methid would be first sort the array and find the mid element it will take $O(nlogn)$ time

In [30]:
%%time
arr = [5,2,1,6,7,3,1,5,2]
# first method
def median(arr):
    arr2 = []
    while arr:
        largest = max(arr)
        arr2.append(largest)
        arr.remove(largest)

    return arr2
arr = median(arr)
n = len(arr)
Median = n//2
print(arr[Median])

3
CPU times: user 202 µs, sys: 0 ns, total: 202 µs
Wall time: 172 µs


In [34]:
%%time
arr = [5,2,1,6,7,3,1,5,2]
arr = sorted(arr)
n = len(arr)
Median = n//2
print(arr[Median])

3
CPU times: user 959 µs, sys: 0 ns, total: 959 µs
Wall time: 978 µs


In [31]:
%%time
import numpy as np
arr = [5,2,1,6,7,3,1,5,2]
print(np.median(arr))

3.0
CPU times: user 2.39 ms, sys: 0 ns, total: 2.39 ms
Wall time: 5.62 ms


So numpy ans is the slowest answer  
apparently first method ans is fastest  
second method is in between numpy and mannual



**4.What is the gradient of the following function with respect to x and y?
$x^2y + y^3sin(x)$**


**ANSWER**
The derivative of this function w.r.t x is $2xy + y^3cos(x)$

The derivative of this function w.r.t y is $x^2 + 3y^2sin(x)$

the formula for  $gradf(x,y)$ = $(df/dx) i + (df/dy) j$

hence $gradf(x^2y+y^3sin(x))$ is $(2xy+y^3cos(x))i + (x^2+3y^2sin(x))j$

**5.Use $JAX$ to confirm the gradient evaluated by your method matches the analytical solution corresponding to a few random values of x and y.**

**ANSWER**

let putting the value x = 2 and y = 3 in my ans

putting them in grad wrt to x is 0.76

putting them in grad wrt to y is 28.5

In [26]:
import jax
import jax.numpy as jnp
def f(x, y):
    return (x**2)*y + y**3 * jnp.sin(x)

# Compute the gradient of f with respect to x
grad_f_x = jax.grad(f, argnums=0)

# Compute the gradient of f with respect to y
grad_f_y = jax.grad(f, argnums=1)

x, y = 2.0,3.0
gradient_x = grad_f_x(x, y)
gradient_y = grad_f_y(x, y)

print(f"The gradient of f with respect to x at x=2 and y=3 is {gradient_x}")
print(f"The gradient of f with respect to y at x=2 and y=3 is {gradient_y}")

The gradient of f with respect to x at x=2 and y=3 is 0.7640352249145508
The gradient of f with respect to y at x=2 and y=3 is 28.551029205322266


Hence my ans and function ans is same

**6.Use sympy to confirm that you obtain the same gradient analytically.**

**7.Create a Python nested dictionary to represent hierarchical information. We want to store record of students and their marks. Something like:**

**2022
Branch 1
Roll Number: 1, Name: N, Marks:
Maths: 100, English: 70 …
Branch 2
2023
Branch 1
Branch 2
2024
Branch 1
Branch 2
2025
Branch 1
Branch 2**

**8.Store the same information using Python classes. We have an overall database which is a list of year objects. Each year contains a list of branches. Each branch contains a list of students. Each student has some properties like name, roll number and has marks in some subjects.**

**9.Using matplotlib plot the following functions on the domain: x = 0.5 to 100.0 in steps of 0.5.**





**10.Using numpy generate a matrix of size 20X5 containing random numbers drawn uniformly from the range of 1 to 2. Using Pandas create a dataframe out of this matrix. Name the columns of the dataframe as “a”, “b”, “c”, “d”, “e”. Find the column with the highest standard deviation. Find the row with the lowest mean.**

In [50]:
import numpy as np
import pandas as pd

matrix = np.random.uniform(low=1, high=2, size=(20, 5))
df = pd.DataFrame(matrix, columns=["a", "b", "c", "d", "e"])

stdlist = df.std()
maxstd = stdlist.idxmax()

row_means = df.mean(axis=1)
min_mean_row = row_means.idxmin()
# Print the results
print("DataFrame:")
print(df)
print(f"\nColumn with the highest standard deviation:{maxstd}")
print(f"\nRow with the lowest mean:{min_mean_row}")

DataFrame:
           a         b         c         d         e
0   1.926112  1.366391  1.778560  1.808073  1.054215
1   1.355747  1.346558  1.235920  1.621997  1.353284
2   1.031031  1.774050  1.168125  1.757015  1.449407
3   1.967964  1.226588  1.098450  1.826845  1.619901
4   1.672545  1.018833  1.640804  1.998648  1.047693
5   1.234175  1.316735  1.526321  1.645426  1.339785
6   1.097335  1.056944  1.872183  1.561860  1.888486
7   1.776724  1.514923  1.404096  1.160741  1.446838
8   1.998340  1.284703  1.738122  1.366969  1.360575
9   1.432811  1.223420  1.742888  1.173282  1.135794
10  1.844908  1.482024  1.649849  1.837192  1.519264
11  1.539662  1.506776  1.295621  1.963304  1.243578
12  1.841972  1.371436  1.367595  1.544713  1.507227
13  1.498538  1.613712  1.466394  1.914980  1.069973
14  1.537855  1.593219  1.816286  1.668931  1.320352
15  1.831152  1.521798  1.808021  1.343073  1.996693
16  1.170457  1.146687  1.070543  1.968126  1.344049
17  1.516825  1.396450  1.902974  1

**11.Add a new column to the dataframe called “f” which is the sum of the columns “a”, “b”, “c”, “d”, “e”. Create another column called “g”. The value in the column “g” should be “LT8” if the value in the column “f” is less than 8 and “GT8” otherwise. Find the number of rows in the dataframe where the value in the column “g” is “LT8”. Find the standard deviation of the column “f” for the rows where the value in the column “g” is “LT8” and “GT8” respectively.**

**12.Write a small piece of code to explain broadcasting in numpy.**


In [58]:
#Broadcasting in numpy is a function using which we can operate between two arrays of non similar dimensions
arr = np.array([[5,6,2,1],[8,6,4,2]])
arr2 = np.array([2,3,4,8])

arr3 = arr + arr2
# Broadcasting replicates therow in arr2
print(arr)
print("\n",arr2)
print("\n",arr3)

[[5 6 2 1]
 [8 6 4 2]]

 [2 3 4 8]

 [[ 7  9  6  9]
 [10  9  8 10]]


**13.Write a function to compute the argmin of a numpy array. The function should take a numpy array as input and return the index of the minimum element. You can use the np.argmin function to verify your solution.**


In [48]:
arr = [5,2,51,66,7,3,1,5,2]

def Argmin(arr):
  ind = 0
  for i in range(len(arr)):
    if(arr[ind] > arr[i]):
      ind = i
  return ind
index = Argmin(arr)
print(index)

6
CPU times: user 1.54 ms, sys: 0 ns, total: 1.54 ms
Wall time: 1.59 ms


In [49]:
x=np.argmin(arr)
print(x)

6


Hence My ans is correct

